api/services/coinsService.js

```diff
@@ -0,0 +1,232 @@
+      resultMap[missingPid] = data;
+    }
+    updateLocalCache(localMetadataCache);
+  }
+
+  for (const p of pidArray) {
+    if (!resultMap[p]) {
+      resultMap[p] = null;
+    }
+  }
+
+  return resultMap;
+}
+
+/**
+ * Retrieve the current coin data from Redis. 
+ * If `withMetadata=true`, attach metadata using the same logic as getCoinMetadata.
+ */
+async function getCoinsService(pidString, options = {}) {
+  const { withTTL = false, withMetadata = false } = options;
+
+  const pids = pidString
+    .split(',')
+    .map(p => p.trim().toLowerCase())
+    .filter(p => p);
+
+  if (!pids.length) {
+    throw new Error("The 'pid' query parameter must contain at least one valid pid.");
+  }
+
+  const redisKeys = pids.map(pid => `price_${pid}`);
+  const coinsData = await getMultipleKeyDetails(redisKeys, withTTL);
+
+  let metadataMap = {};
+  if (withMetadata) {
+    metadataMap = await getCoinMetadata({ pid: pids.join(',') });
+  }
+
+  const coins = pids.map(pid => {
+    const redisKey = `price_${pid}`;
+    const data = coinsData[redisKey] || { value: {}, ttl: null };
+    const { pid: storedPid, price, confidence, source } = data.value;
+
+    return {
+      pid: storedPid || pid,
+      price,
+      confidence,
+      source,
+      ttl: withTTL ? data.ttl : undefined,
+      metadata: withMetadata ? (metadataMap[pid] || null) : undefined
+    };
+  });
+
+  return { coins };
+}
+
+async function findClosestDocForPid(client, pid, timestampMs) {
+  const afterQuery = {
+    index: 'coins-timeseries-*',
+    size: 1,
+    body: {
+      query: {
+        bool: {
+          must: [
+            { term: { pid } },
+            { range: { ts: { gte: timestampMs } } }
+          ]
+        }
+      },
+      sort: [{ ts: { order: 'asc' } }]
+    }
+  };
+
+  const beforeQuery = {
+    index: 'coins-timeseries-*',
+    size: 1,
+    body: {
+      query: {
+        bool: {
+          must: [
+            { term: { pid } },
+            { range: { ts: { lte: timestampMs } } }
+          ]
+        }
+      },
+      sort: [{ ts: { order: 'desc' } }]
+    }
+  };
+
+  const [afterResp, beforeResp] = await Promise.all([
+    client.search(afterQuery),
+    client.search(beforeQuery)
+  ]);
+
+  const afterHit = afterResp.hits?.hits[0]?._source || null;
+  const beforeHit = beforeResp.hits?.hits[0]?._source || null;
+
+  if (!afterHit && !beforeHit) return null;
+  if (afterHit && !beforeHit) return afterHit;
+  if (!afterHit && beforeHit) return beforeHit;
+
+  const diffAfter = Math.abs(afterHit.ts - timestampMs);
+  const diffBefore = Math.abs(beforeHit.ts - timestampMs);
+  return diffAfter <= diffBefore ? afterHit : beforeHit;
+}
+
+/**
+ * Retrieve timeseries data:
+ * - If `timestamp` is provided, for each pid we return the doc whose ts is closest.
+ * - Otherwise, we fetch all docs (or a range if startDate/endDate are given).
+ */
+async function getCoinsTimeseries({ pid, startDate, endDate, timestamp }) {
```

Comment:
also, need another param: time scale, like 1m/5m, 30m/1h/1d etc, again need to check how current implementation handles this